pkg/querier/block_store.go

```diff
@@ -0,0 +1,237 @@
+package querier
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/alecthomas/units"
+	"github.com/cortexproject/cortex/pkg/ingester"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/thanos-io/thanos/pkg/model"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/objstore/s3"
+	"github.com/thanos-io/thanos/pkg/store"
+	storecache "github.com/thanos-io/thanos/pkg/store/cache"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// UserStore is a multi-tenant version of Thanos BucketStore
+type UserStore struct {
+	logger log.Logger
+	cfg    s3.Config
+	bucket objstore.BucketReader
+	stores map[string]*store.BucketStore
+	client storepb.StoreClient
+}
+
+// NewUserStore returns a new UserStore
+func NewUserStore(logger log.Logger, s3cfg s3.Config) (*UserStore, error) {
+	bkt, err := s3.NewBucketWithConfig(logger, s3cfg, "cortex-userstore")
+	if err != nil {
+		return nil, err
+	}
+
+	u := &UserStore{
+		logger: logger,
+		cfg:    s3cfg,
+		bucket: bkt,
+		stores: make(map[string]*store.BucketStore),
+	}
+
+	serv := grpc.NewServer()
+	storepb.RegisterStoreServer(serv, u)
+	l, err := net.Listen("tcp", "")
+	if err != nil {
+		return nil, err
+	}
+	go serv.Serve(l)
+
+	cc, err := grpc.Dial(l.Addr().String(), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	u.client = storepb.NewStoreClient(cc)
+
+	return u, nil
+}
+
+// InitialSync iterates over the s3 bucket creating user bucket stores, and calling InitialSync on each of them
+func (u *UserStore) InitialSync(ctx context.Context) error {
+	if err := u.syncUserStores(ctx, func(ctx context.Context, s *store.BucketStore) error {
+		return s.InitialSync(ctx)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SyncStores iterates over the s3 bucket creating user bucket stores
+func (u *UserStore) SyncStores(ctx context.Context) error {
+	if err := u.syncUserStores(ctx, func(ctx context.Context, s *store.BucketStore) error {
+		return s.SyncBlocks(ctx)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *UserStore) syncUserStores(ctx context.Context, f func(context.Context, *store.BucketStore) error) error {
```

Comment:
We'll hit this quite quickly in a multi-tenant context: https://github.com/thanos-io/thanos/issues/1471


Further, I don't see a good way to horizontally scale this as there is no sharding and all queriers should have all the blocks.

Reply:
Yea make sense. I wonder if we could do something like sharding users to a stateful set of queriers, so only some querier have a given user. This problem is definitely food for thought.

Reply:
Let me add some math. Scenario:
- 1K tenants (reasonably low)
- 14 months retention (approx. 31 days / month)
- 2 hours TSDB block range

Total number of blocks per tenant: 14 * 31 * (24 / 2) = 5.2M blocks
Total number of API calls required to list all the blocks (assuming 1000 max items in the response): 5.2M / 1000 = 5200

The initial sync of a querier needs to download the `meta.json` file for each block for each tenant, which means 5.2M calls to the object store, which looks not feasible to me.

A possible alternative would be index blocks by time range and implement it in a fully lazy way. We may prepend each block name stored on S3 with the min timestamp of that block (using a timestamp format which guarantees alphabetical ordering).

Given a query for the time range [A,B], we need to sync only the blocks whose min timestamp is [A-delta,B] where `delta` is the max block range allowed by the system. The querier may have a config options to initially sync only the last X hours of blocks (assuming most queries are against recent data), while older blocks metadata will be lazily fetched.

Moreover, the periodic sync to find new blocks wouldn't require to iterate over all blocks of a tenant, but we could set the "StartAfter" param (of S3 ListBucket API) to the last min timestamp synched - a delta.

Does it make any sense?

Reply:
S3-based indexing is just a trick to avoid introducing another dependency. Indexes on Bigtable (like it work for Cortex chunks today) would serve the same purpose in a probably cleaner way.

Reply:
Totally makes sense. I do personally think this is still an optimization for after merging this PR, as this change is already large. Blocks are already sorted by ULID so we would have time ordering already so I think we could already lazy sync on that.

Reply:
Do you view this as a blocker for the initial feature merge?

Reply:
.. which is discussed here: https://github.com/thanos-io/thanos/issues/1318 Please join the discussion if you have ideas that will work for you as well! (: